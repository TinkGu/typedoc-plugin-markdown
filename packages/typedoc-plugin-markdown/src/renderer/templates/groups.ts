import {
  DeclarationReflection,
  ReflectionCategory,
  ReflectionGroup,
} from 'typedoc/dist/lib/models';

import { MarkdownBuilder } from '../tools/builder';
import { heading, horizontalRule } from '../tools/elements';
import { memberTemplate } from './member';

export function groupsTemplate(groups: ReflectionGroup[]) {
  const md = new MarkdownBuilder();

  groups
    ?.filter((group) => !group.allChildrenHaveOwnDocument())
    .forEach((group) => {
      if (group.categories) {
        group.categories
          ?.filter((category) => !category.allChildrenHaveOwnDocument())
          .forEach((category) => {
            md.add(heading(2, `${group.title} ${category.title}`));
            md.add(getMembers(category));
          });
      } else {
        md.add(heading(2, group.title));
        md.add(getMembers(group));
      }
    });

  return md.print();
}

function getMembers(item: ReflectionGroup | ReflectionCategory) {
  const md = new MarkdownBuilder();
  item.children.forEach((child, index) => {
    md.add(memberTemplate(child as DeclarationReflection));
    if (index !== item.children.length - 1) {
      md.add(horizontalRule());
    }
  });
  return md.print();
}
