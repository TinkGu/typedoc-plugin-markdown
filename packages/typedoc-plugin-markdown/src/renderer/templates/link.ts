import * as path from 'path';

import settings from '../settings';
import { link } from '../tools/elements';
import { escapeChars } from '../tools/utils';

export function linkTemplate(label: string, to = '', escapeLabel = true) {
  label = escapeLabel ? escapeChars(label) : label;
  return link(label, getUrl(to));
}

function getUrl(to: string) {
  const { options, activeUrl } = settings;

  const urlPrefix = /^(http|ftp)s?:\/\//;
  if (urlPrefix.test(to)) {
    return to;
  } else {
    return options.publicPath
      ? getPublicUrl(options.publicPath, to)
      : getRelativeUrl(activeUrl, to);
  }
}

function getPublicUrl(publicPath: string, url: string) {
  return publicPath + url;
}

function getRelativeUrl(from: string, to: string) {
  const relative = path.relative(path.dirname(from), path.dirname(to));
  return path.join(relative, path.basename(to)).replace(/\\/g, '/');
}
