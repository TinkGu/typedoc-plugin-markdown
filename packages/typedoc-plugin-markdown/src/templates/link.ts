import * as path from 'path';
import { URL_PREFIX } from '../constants';
import { useSettings } from '../context';
import { link } from '../tools/elements';
import { escapeChars } from '../tools/utils';

export function linkTemplate(label: string, to = '', escapeLabel = true) {
  label = escapeLabel ? escapeChars(label) : label;
  return link(label, getUrl(to));
}

function getUrl(to: string) {
  const { activeUrl, publicPath } = useSettings();

  if (URL_PREFIX.test(to)) {
    return to;
  } else {
    return publicPath
      ? getPublicUrl(publicPath, to)
      : getRelativeUrl(activeUrl, to);
  }
}

function getPublicUrl(publicPath: string, url: string) {
  return publicPath + url;
}

function getRelativeUrl(from: string, to: string) {
  const relative = path.relative(path.dirname(from), path.dirname(to));
  return path.join(relative, path.basename(to)).replace(/\\/g, '/');
}
