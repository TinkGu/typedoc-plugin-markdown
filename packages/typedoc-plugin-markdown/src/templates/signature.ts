import {
  DeclarationReflection,
  ReflectionKind,
  SignatureReflection,
} from 'typedoc';
import { MarkdownBuilder } from '../tools/builder';
import { heading } from '../tools/elements';
import { comment, commentsTemplate } from './comments';
import { signatureTitleTemplate } from './signature.title';
import { sourcesTemplate } from './sources';
import { parameterTableTemplate } from './table.parameter';
import { propertyTableTemplate } from './table.property';
import { typeParameterTableTemplate } from './table.type-parameter';
import { typeTemplate } from './type';

export function signatureTemplate(
  model: SignatureReflection,
  nested = false,
  accessor?: string,
) {
  const md = new MarkdownBuilder();

  const typeDeclaration = (model.type as any)
    ?.declaration as DeclarationReflection;

  md.add(signatureTitleTemplate(model, accessor));

  if (model.comment) {
    md.add(commentsTemplate(model.comment));
  }

  if (model.typeParameters?.length) {
    md.add(heading(nested ? 5 : 4, 'Type parameters'));
    md.add(typeParameterTableTemplate(model.typeParameters));
  }

  if (model.parameters?.length) {
    md.add(heading(nested ? 5 : 4, 'Parameters'));
    md.add(parameterTableTemplate(model.parameters));
  }

  if (model.type && !model.parent?.kindOf(ReflectionKind.Constructor)) {
    if (model.type) {
      md.add(heading(nested ? 5 : 4, 'Returns'));
      md.add(typeTemplate(model.type, 'all'));
      if (model.comment?.returns) {
        md.add(comment(model.comment.returns));
      }
    }

    if (typeDeclaration?.signatures) {
      typeDeclaration.signatures.forEach((signature) => {
        md.add(signatureTemplate(signature, true));
      });
    }

    if (typeDeclaration?.children) {
      md.add(propertyTableTemplate(typeDeclaration.children));
    }
  }

  if (!nested) {
    md.add(sourcesTemplate(model));
  }

  return md.print();
}
