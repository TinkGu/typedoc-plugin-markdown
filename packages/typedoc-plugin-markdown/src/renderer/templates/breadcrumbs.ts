import { Reflection } from 'typedoc';
import { PageEvent } from 'typedoc/dist/lib/output/events';
import settings from '../settings';
import { MarkdownBuilder } from '../tools/builder';
import { escapeChars } from '../tools/utils';
import { linkTemplate } from './link';

export function breadcrumbsTemplate(page: PageEvent) {
  const { activeUrl, globalsName, globalsFile, options } = settings;

  const md = new MarkdownBuilder();

  md.add(
    page.url === options.entryDocument
      ? page.project.name
      : linkTemplate(page.project.name, options.entryDocument),
  );

  if (!options.readme.endsWith('none')) {
    if (activeUrl === globalsFile) {
      md.add(globalsName);
    } else {
      md.add(linkTemplate(globalsName, globalsFile));
    }
  }

  const addBreadcrumb = (model: Reflection, md: MarkdownBuilder) => {
    if (model && model.parent) {
      addBreadcrumb(model.parent, md);

      if (activeUrl === model.url) {
        md.add(escapeChars(model.name));
      } else {
        md.add(linkTemplate(model.name, model.url));
      }
    }
  };

  addBreadcrumb(page.model, md);

  return md.print(' / ');
}
