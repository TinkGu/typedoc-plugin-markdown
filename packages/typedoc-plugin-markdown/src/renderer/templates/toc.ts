import {
  DeclarationReflection,
  ProjectReflection,
  ReflectionCategory,
  ReflectionGroup,
} from 'typedoc/dist/lib/models';

import settings from '../settings';
import { MarkdownBuilder } from '../tools/builder';
import { heading, unorderedList } from '../tools/elements';
import { linkTemplate } from './link';

export function tocTemplate(model: ProjectReflection | DeclarationReflection) {
  const options = settings.options;
  const md = new MarkdownBuilder();
  const isVisible = model.groups?.some((group) =>
    group.allChildrenHaveOwnDocument(),
  );
  if ((!this.hideInPageTOC && model.groups) || (isVisible && model.groups)) {
    if (!options.hideInPageTOC) {
      md.add(heading(2, 'Table of contents'));
    }
    model.groups
      ?.filter(
        (group) => !options.hideInPageTOC || group.allChildrenHaveOwnDocument(),
      )
      .forEach((group) => {
        if (group.categories) {
          group.categories.forEach((category) => {
            md.add(getGroup(category, category.title + ' ' + group.title));
          });
        } else {
          md.add(getGroup(group, group.title));
        }
      });
  }

  return md.print();
}

function getGroup(group: ReflectionGroup | ReflectionCategory, title: string) {
  const options = settings.options;
  const md = new MarkdownBuilder();
  md.add(heading(options.hideInPageTOC ? 2 : 3, title));
  const groupItems = group.children.map((child) =>
    linkTemplate(child.name, child.url),
  );
  md.add(unorderedList(groupItems));
  return md.print();
}
