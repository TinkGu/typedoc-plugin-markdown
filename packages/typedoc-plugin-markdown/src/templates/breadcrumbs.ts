import { Reflection, PageEvent, ContainerReflection } from 'typedoc';
import { useSettings } from '../context';
import { MarkdownBuilder } from '../tools/builder';
import { escapeChars } from '../tools/utils';
import { linkTemplate } from './link';

export function breadcrumbsTemplate(page: PageEvent<ContainerReflection>) {
  const { activeUrl, globalsName, globalsFile, entryDocument, readme } =
    useSettings();

  const md = new MarkdownBuilder();

  md.add(
    page.url === entryDocument
      ? page.project.name
      : linkTemplate(page.project.name, entryDocument),
  );

  if (!readme.endsWith('none')) {
    if (activeUrl === globalsFile) {
      md.add(globalsName);
    } else {
      md.add(linkTemplate(globalsName, globalsFile));
    }
  }

  const addBreadcrumb = (model: Reflection, md: MarkdownBuilder) => {
    if (model && model.parent) {
      addBreadcrumb(model.parent, md);

      if (activeUrl === model.url) {
        md.add(escapeChars(model.name));
      } else {
        md.add(linkTemplate(model.name, model.url));
      }
    }
  };

  addBreadcrumb(page.model, md);

  return md.print(' / ');
}
