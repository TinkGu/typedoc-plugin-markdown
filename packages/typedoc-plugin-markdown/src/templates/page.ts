import { DeclarationReflection } from 'typedoc';
import { PageEvent } from 'typedoc/dist/lib/output/events';
import { getContext } from '../context';
import { MarkdownBuilder } from '../tools/builder';
import { heading } from '../tools/elements';
import { breadcrumbsTemplate } from './breadcrumbs';
import { groupsTemplate } from './groups';
import { pageTitleTemplate } from './page-title';
import { reflectionTemplate } from './reflection';
import { tocTemplate } from './toc';

export function pageTemplate(page: PageEvent) {
  const md = new MarkdownBuilder();

  const { hideBreadcrumbs, hidePageTitle } = getContext();

  if (!hideBreadcrumbs) {
    md.add(breadcrumbsTemplate(page));
  }

  if (!hidePageTitle) {
    md.add(heading(1, pageTitleTemplate(page)));
  }

  if (page.model instanceof DeclarationReflection) {
    md.add(reflectionTemplate(page.model));
  } else {
    md.add(tocTemplate(page.model));
    md.add(groupsTemplate(page.model.groups));
  }

  return md.print();
}
