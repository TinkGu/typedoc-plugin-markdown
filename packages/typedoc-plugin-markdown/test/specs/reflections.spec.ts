import { PageEvent } from 'typedoc/dist/lib/output/events';
import * as breadcrumbsTemplate from '../../src/templates/breadcrumbs';
import * as groupsTemplate from '../../src/templates/groups';
import * as linkTemplate from '../../src/templates/link';
import { pageTemplate } from '../../src/templates/page';
import * as signatureTemplate from '../../src/templates/signature';
import * as tocTemplate from '../../src/templates/toc';
import { TestApp } from '../test-app';

describe(`Reflections:`, () => {
  beforeAll(() => {
    jest
      .spyOn(breadcrumbsTemplate, 'breadcrumbsTemplate')
      .mockReturnValue('[BREADCRUMBS]');
    jest
      .spyOn(signatureTemplate, 'signatureTemplate')
      .mockReturnValue('[SIGNATURE]');
    jest.spyOn(tocTemplate, 'tocTemplate').mockReturnValue('[TOC]');
    jest.spyOn(linkTemplate, 'linkTemplate').mockReturnValue('[LINK]');
    jest.spyOn(groupsTemplate, 'groupsTemplate').mockReturnValue('[GROUPS]');
  });

  afterAll(() => {
    jest.restoreAllMocks();
  });
  describe(`(header)`, () => {
    let testApp: TestApp;

    beforeAll(() => {
      testApp = new TestApp(['reflections.ts']);
      testApp.bootstrap({
        hideBreadcrumbs: false,
        hidePageTitle: true,
      });
    });

    afterAll(() => {
      testApp.cleanup();
    });

    test(`should compile template with breadcrumbs and without title`, () => {
      expect(
        pageTemplate({
          model: testApp.project.children[0],
          project: testApp.project,
        } as PageEvent),
      ).toMatchSnapshot();
    });
  });

  describe(`(template)`, () => {
    let testApp: TestApp;
    beforeAll(() => {
      testApp = new TestApp(['reflections.ts']);
      testApp.bootstrap({
        hideBreadcrumbs: true,
        hidePageTitle: false,
      });
    });

    afterAll(() => {
      testApp.cleanup();
    });

    test(`should compile module without breadcrumbs and with project title`, () => {
      expect(
        pageTemplate({
          model: testApp.project.children[0],
          project: testApp.project,
        } as PageEvent),
      ).toMatchSnapshot();
    });

    test(`should compile a callable reflection`, () => {
      expect(
        pageTemplate({
          model: testApp.findReflection('CallableReflection'),
          project: testApp.project,
        } as PageEvent),
      ).toMatchSnapshot();
    });

    test(`should compile an indexable reflection`, () => {
      expect(
        pageTemplate({
          model: testApp.findReflection('IndexableReflection'),
          project: testApp.project,
        } as PageEvent),
      ).toMatchSnapshot();
    });

    test(`should compile implemented class`, () => {
      expect(
        pageTemplate({
          model: testApp.findReflection('ImplementedClass'),
          project: testApp.project,
        } as PageEvent),
      ).toMatchSnapshot();
    });
  });
});
