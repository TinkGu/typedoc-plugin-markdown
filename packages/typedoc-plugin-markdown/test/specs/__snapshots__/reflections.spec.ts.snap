// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reflections: (header) should compile template with breadcrumbs and without title 1`] = `
"[BREADCRUMBS]

[TOC]

[GROUPS]"
`;

exports[`Reflections: (template) should compile a callable reflection 1`] = `
"# Interface: CallableReflection

## Callable

### CallableReflection

[SIGNATURE]

[TOC]"
`;

exports[`Reflections: (template) should compile an indexable reflection 1`] = `
"# Interface: IndexableReflection

[TOC]"
`;

exports[`Reflections: (template) should compile implemented class 1`] = `
"# Class: ImplementedClass

## Implements

- [LINK]

[TOC]

[GROUPS]"
`;

exports[`Reflections: (template) should compile module without breadcrumbs and with project title 1`] = `
"# Enumeration: EnumReflection

[TOC]

[GROUPS]"
`;
